### DISCUSSION TITLE
DB Schema definition

### DISCUSSION THREAD

**Discussion by J0sueTM** [emerauth/emerauth#11] [General]
Our idea for now of Emerauth's DB Schema works is still not well founded. Our abstract idea for now is:

1. There are users
2. A user can be authenticated through a provider (otp, email, google, etc)
3. These users can be authorized based on their user role
4. A user role has access' and restrictions on applications
5. A user's session should be well logged and traceable

Based on these requirements (please point out any more needed), my thoughts of how all these will go down to a DB comes quite simply by separating the persisted stuff from the non-persisted stuff, which I initially separate as:

**Persisted**:  Users, Session and Action Logs
**Non Persisted**: Providers, User roles and Applications

Also, session and action logs don't necessarily mean being persisted on the same DB the users are being stored at. Of course, if needed, they can, but making it all work through a log appender that goes to, idk Sentry or ElasticSearch, or even though a queue worker like Kafka or Redis, could be under our radar.

What do you guys think? @emerauth/core-developers

**Comment by lanjoni** [comment]
I agree with you. I believe that especially User roles should be controlled at the final application layer. Providers would not necessarily need to be stored, unless multiple links were possible in the same account, right? Example: the same account connecting Google and GitHub.

Furthermore, will the definition of schemas be made and designed using a library like [malli](https://github.com/metosin/malli) for example?

**Reply by J0sueTM** [reply]
Sure! Malli could be a good entry point for verifying the `config.edn`, but the process itself will be done through the `aero` lib, right?

Besides the application level of emerauth (which includes user roles, sessions etc) this is what i envision for our DB schema:
![image](https://github.com/emerauth/emerauth/assets/41875891/346c5c5e-664a-4ab0-954f-0c9b67f6c9cb)

It's simple, but its enough for a working α 0.1.0 release
